---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---


load packages and define folders
```{r}
##install.packages("tidyverse")
library(tidyverse)
data.dir = "/home/haky/IntroStatGen/data/hapmap/"
bin.dir = "/home/haky/IntroStatGen/bin/"
out.dir ="/home/haky/IntroStatGen/out/"
gist.dir = "/home/haky/IntroStatGen/repo-IntroStatGen/gists/"

plinkname = paste0(bin.dir,"plink")
bedheader = paste0(data.dir,"hapmap2-common")

```



```{r}
##read hapmap.bim
bimdata = read_tsv(paste0(bedheader, ".bim"),col_names = FALSE)
names(bimdata) = c("chr","rsid","dis","pos","A1","A2")

##sample prop % of SNPs to make them causal
propC = 0.10
Mt = nrow(bimdata)
MC = round(propC * Mt) ## number of causal SNPs
seednumber = 1

```


Randomly choose 10% SNPs as causal
```{r, eval=FALSE}
set.seed(seednumber) ## we want to get the same SNPs each time we run this
indC = sample(1:Mt, MC) ## select Mc SNPs to be simulated as causal

```

Simulate betas as normally distributed for the 10% of SNPs selected above
```{r simulate betas, eval=FALSE}

set.seed(seednumber + 1)
betaC = rnorm(MC,mean=0, sd=1/sqrt(MC))
hist(betaC)

## generate weights file, Y = sum(beta * X) + error
weightdata = bimdata[indC,] %>% select(rsid,A1) 
weightdata$betas = betaC
weightfile = paste0(out.dir, "weights-seed",seednumber,".txt")
write_tsv(weightdata,path=weightfile)

```

calculate polygenic core using --score option in plink
```{r}

##https://www.cog-genomics.org/plink/1.9/score
comm = paste0(bin.dir,"plink --bfile ", bedheader," --score ", weightfile, " header sum --out ", out.dir,"polyscore-seed",seednumber)
print(comm)
system( comm )

## read sum(beta * X) from polyscore-seed1.profile
PSdata = read_table(paste0(out.dir,"polyscore-seed",seednumber,".profile"))
head(PSdata)
hist(PSdata$SCORESUM)
```

Calculate var of epsilon given variance of beta and heritability
```{r}

h2 = 0.2
sigepsi2 = (1-h2)/h2 * var(PSdata$SCORESUM)

## homework problem: show analytically that variance of epsilon defined above leads to the heritability h2. 

```


Calculate phenotype and write file to be used with plink
```{r}

phenofile = paste0(out.dir,"simpheno-",seednumber,".txt")


nsamp = nrow(PSdata)
set.seed(seednumber + 2); ## use seednumber + 2 to get reproducible simulations
epsilon = rnorm(nsamp,mean=0,sd=sqrt(sigepsi2) ) 
PSdata = PSdata %>% mutate(PHENO = SCORESUM + epsilon)
write_tsv(PSdata %>% select(FID,IID,PHENO,SCORESUM),path=paste0(phenofile))

## homework problem: show empirically that the heritability of the phenotype is = h2
## var(SCORESUM) / var(PHENO) 

```

run GWAS
```{r }
outheader = paste0(out.dir,"simpheno-",seednumber)

comm = paste0(plinkname, " --bfile ", bedheader, " --assoc --pheno ",phenofile, " --pheno-name PHENO ", " --out ", outheader)
print(comm)
system(comm)
```


plots histogram of p values
```{r}

## read plink association results in simpheno-1.qassoc

qassoc = read_table(paste0(outheader, ".qassoc"))
dim(qassoc)
hist(qassoc$P)
source(paste0(gist.dir, "qqunif.r")) ## load qqunif function from the gist folder
qqunif.plot(qassoc$P,main = paste0("sim with h2 = ",round(h2,2)) )


```


PCA calculation using plink
```{r}
## generate PCs using plink ## this takes some time
if(F)
{comm = paste0(plinkname, " --bfile ", bedheader, " --pca ", " --out ", outheader)
print(comm)
system(comm)
}

## read plink calculated PCs
pcplink = read_table2(paste0(outheader,".eigenvec"),col_names = FALSE)
dim(pcplink)
head(pcplink)
names(pcplink) = c("FID","IID",paste0("PC", c(1:(ncol(pcplink)-2))) )
popinfo = read_tsv(paste0(data.dir,"relationships_w_pops_051208.txt"))

pcplink = popinfo %>% inner_join(pcplink, by=c("FID"="FID", "IID"="IID"))
## plot PC1 vs PC2
pcplink %>% ggplot(aes(PC1,PC2,col=population)) + geom_point()

```

runnig GWAS adjusting for PCs
```{r}

## takes several minutes to run
if(F)
{covarfile = paste0(outheader,".eigenvec")
comm = paste0(plinkname, " --bfile ", bedheader, " --linear --pheno ",phenofile, " --pheno-name PHENO --covar ", covarfile, " --covar-number 3-7", " --out ", outheader)
print(comm)
system(comm)
## keep only header and ADD rows from plink output 
system(paste0("grep ADD ", outheader,".assoc.linear > ", outheader, ".assoc.linear.addonly") )
}

## read plink association results
assocPC = read_table(paste0(outheader, ".assoc.linear.addonly"),col_names=FALSE)
tempo = read_table(paste0(outheader, ".assoc.linear"),n_max=1)
names(assocPC) = names(tempo)

hist(assocPC$P)
qqunif.plot(assocPC$P)

## adjusting for PC's p values follow the null distribution more closely.

```
Not much difference without adjusting for PC
```{r}

identical(assocPC$SNP,qassoc$SNP) ## if this is not TRUE, need to use inner_join to match SNPs
ind=sample(1:nrow(assocPC), 1e4); plot(-log10(assocPC$P[ind]), -log10(qassoc$P[ind]) )
ind=sample(1:nrow(assocPC), 1e4); plot(assocPC$BETA[ind], qassoc$BETA[ind] )

```


compare true betas with estimates
Notice 90% of betas are 0
```{r, echo=FALSE}
install.packages("qqman")
library(qqman)
ind=sample(1:nrow(assocPC), 1e5);manhattan(qassoc[ind,])

```



calculate h2
```{r}

## calculate GRM
gctafile = paste0(bin.dir,"gcta64")
comm = paste0(gctafile, " --bfile ", bedheader," --make-grm-bin --out ", outheader)
print(comm)
system(comm)

```



calculate h2left using ch1 to ch10 

calculate h2right using ch1 to ch10 

use Yang formula to adjust


```{r}

```

